### PR TITLE
Normalize placeholder names in AOTAutogradCache

### LINKED ISSUE #157792: Consider changing AOTAutograd cache to hit on graphs with different input and node names
We ran into this in vLLM, where vLLM:
1) creates one big graph with a lot of transformer layers
2) splits the big graph into multiple graphs, each of which are the same
3) compiles each graph separately (sending it through AOTAutograd). All of these graphs cache miss due to having different input names.

The workaround I am currently applying is to normalize the inputs of the graphs before sending them through AOTAutograd. It's not clear to me if the right long-term solution is:
1) AOTAutograd cache becomes agnostic to names
2) user is supposed to do some sort of hierarchical compilation
3) user is supposed to know these quirks with AOTAutograd cache and program with those in mind

### PR DESCRIPTION
**Pull request by jamesjwu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157916

This PR adds a pass to sanitize_gm_for_cache which normalizes all placeholder names across input dynamo graphs to AOTAutogradCache. This is safe because nothing underneath AOTAutograd uses the node names on the
original dynamo graph: AOTAutograd re-traces with its own nodes, and guards are
in terms of original sources rather than placeholder names.

Note that the dynamo output graphs traced by tlparse will not show this change because it's done before this sanitization step. The aot autograd outputs also will not change because AOTAutograd's own traced graphs don't use the original placeholders of the dynamo graph. Thus, this change is essentially a no-op from everyone's perspective except for cache key checks.

Fixes #157792 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 1]
```diff
@@ -380,6 +381,52 @@ def _reduce_tensor(self, tensor):
         return (_ident, (metadata,))
 
 
+@contextlib.contextmanager
+def normalize_placeholder_names(gm: torch.fx.GraphModule):
```
Comment (jamesjwu) [inline torch/_functorch/_aot_autograd/autograd_cache.py:388]: FWIW idk if this is the best way to implement this function: I had to hack in a few extra changes to each node like changing the target and used_names in the namespace so that we properly reconstruct the dynamo graph each time. 

I wish it was easier than this, but I can't think of an easier way that isn't considerably slower, i.e. copying the entire graph. Tests seem to pass with this approach, though.

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 2]
```diff
@@ -380,6 +381,52 @@ def _reduce_tensor(self, tensor):
...
+    that are isomorphic when normalizing names can hit the same cache entry.
+
+    This is safe because nothing underneath AOTAutograd uses the node names on the
+    original dynamo graph: AOTAutograd re-traces with its own nodes, and guards are
+    in terms of original sources rather than placeholder names.
+    """
+    old_placeholder_names = [
+        str(n.target)
+        for n in gm.graph.nodes
+        if n.op == "placeholder" and n.type != torch.SymInt
```
Comment (jamesjwu) [inline torch/_functorch/_aot_autograd/autograd_cache.py]: @bobrenjc93 I'm pretty sure this isn't the way to check if the graph node is symbolic. Is there a preferred way? I don't want to rename the symbolic placeholders because symbolic hashing already makes sure they're consistent.

File: test/dynamo/test_backward_higher_order_ops.py [hunk 3]
```diff
@@ -131,14 +131,11 @@ def fn(x, y):
                     """\
 class GraphModule(torch.nn.Module):
     def forward(self, L_inputs_ : list, s69: "Sym(s21)", L_sizes_0_: "f32[0, s21]"):
-        l_inputs_ = L_inputs_
```
Comment (jamesjwu) [inline test/dynamo/test_backward_higher_order_ops.py]: cc @anijain2305 , this is what I mean: the current implementation of normalize_placeholder_for_gm() will remove some of these unnecessary local variable name changes. I can't figure out a clean way to preserve these (from what I can tell, unnecessary) name changes

**Comment by jamesjwu** [comment]
I bit the bullet and decided to just make a copy of the graph module instead of the hacky thing I was doing before. This way I can make edits to the new gm without afffecting the old one at all, which greatly simplifies the code. The cost here is an extra copy, but until I see evidence that this is going to affect compile times I think it's fine. 

(For low overhead scenarios like precompile, we're not calculating any cache keys anyway)

**Comment by jamesjwu** [comment]
Ok I figured out a cleaner way to do it that does not modify the graph: we just have to be careful about preserving the old namespace when renaming nodes. I think this new context manager is idempotent and also cleaner, while not requiring us to clone the entire graph. 

It obviously still needs to do some weird things like preserving the old used_names set, but I think it's much more reasonable now.

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 4]
```diff
@@ -380,6 +380,57 @@ def _reduce_tensor(self, tensor):
...
+    # as-is
+    if not hasattr(gm, "graph"):
+        yield
+        return
+
+    # Track all the old state of placeholders
+    old_placeholder_names = []
+    old_used_names = copy(gm.graph._graph_namespace._used_names)
+    i = 0
+    for n in gm.graph.find_nodes(op="placeholder", sort=True):
```
Comment (zou3519) [inline torch/_functorch/_aot_autograd/autograd_cache.py:407]: will this noticeably impact perf for models that have a lot of parameters?

Comment (zou3519) [reply torch/_functorch/_aot_autograd/autograd_cache.py:407]: If this is too expensive then we are gonna need to do some tradeoff analysis. If vLLM is the only situation that benefits from this, vLLM can just manually do the input name normalization if needed.

Comment (jamesjwu) [reply torch/_functorch/_aot_autograd/autograd_cache.py:407]: I'll kick off a benchmark to confirm, but I feel like the cost of looking for placeholder nodes is relatively cheap.

**Comment by jamesjwu** [comment]
https://hud.pytorch.org/benchmark/huggingface/inductor_dynamic?dashboard=torchinductor&startTime=Sun%2C%2006%20Jul%202025%2019%3A34%3A38%20GMT&stopTime=Sun%2C%2013%20Jul%202025%2019%3A34%3A38%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(h100)&lBranch=gh/jamesjwu/172/head&lCommit=cad3701c0899ba21e53150c5ae5db875868049e7&rBranch=gh/jamesjwu/172/base&rCommit=1308093ce0b0915bac65906f6207fbc8969f97a6

Seems fine on our OSS benchmarks. I will land it with is_fbcode() false for now, so I can test it internally as well. I'm having some trouble with ghimport. For VLLM, they can set that config flag on for now until it's on by default internally.

File: torch/_functorch/config.py [hunk 5]
```diff
@@ -61,6 +61,10 @@
 # need to add env vars or make it configurable
 bundled_autograd_cache: bool = False
 
+# Whether or not to normalize placeholder names in graphs
+# from dynaom in AOTAutogradCache
```
Comment (zou3519) [inline torch/_functorch/config.py:65]: nit: dynamo

**Review by zou3519 — APPROVED** [review]
